build/commands/lib/getAffectedTests.js

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) 2025 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+const { promisify } = require('util')
+const { readFile, writeFile } = require('fs/promises')
+const exec = promisify(require('child_process').execFile)
+const path = require('path')
+const config = require('./config')
+const { unlink } = require('fs-extra')
+const { randomUUID } = require('crypto')
+const { getApplicableFilters, getTestsToRun } = require('./testUtils')
+const { tmpdir } = require('os')
+
+const getTestTargets = (outDir, filters = ['//*']) => {
+  const { env, shell } = config.defaultOptions
+  return exec(
+    'gn',
+    ['ls', outDir, '--type=executable', '--testonly=true', ...filters],
+    { env, shell },
+  ).then((x) => x.stdout.trim().split('\n'))
+}
+
+const asGnTarget = (file) =>
+  ('//brave/' + file).replace('brave/chromium_src/', '')
+
+async function getModifiedFiles(target = 'HEAD~', base = null) {
+  const args = ['diff', '--name-only', target, base].filter((x) => x)
+  const maxBuffer = 1024 * 1024 * 5
+  return exec('git', args, { maxBuffer }).then((x) =>
+    x.stdout
+      .trim()
+      .split('\n')
+      .filter((x) => x),
+  )
+}
+
+async function getReferenceCommit() {
+  // JENKINS sets GIT_PREVIOUS_SUCCESSFUL_COMMIT
+  if (process.env.GIT_PREVIOUS_SUCCESSFUL_COMMIT) {
+    return process.env.GIT_PREVIOUS_SUCCESSFUL_COMMIT
+  }
+
+  const currentBranch = await exec('git', [
+    'rev-parse',
+    '--abbrev-ref',
+    'HEAD',
+  ]).then((x) => x.stdout)
+
+  if (currentBranch !== 'master') {
+    return 'origin/master'
+  }
+
+  // compare with previous commit
+  return 'HEAD~'
+}
+
+async function analyzeAffectedTests(
+  outDir,
+  { filters = ['//*'], files = [], since = null } = {},
+) {
+  const targetCommit =
+    !since || since === true ? await getReferenceCommit() : since
+
+  const root = path.resolve(process.cwd(), '../')
+  outDir = path.isAbsolute(outDir) ? outDir : `${root}/${outDir}`
+  const testTargets = await getTestTargets(outDir, filters)
+  const modifiedFiles = [
+    ...(await getModifiedFiles(targetCommit)),
+    ...files,
+  ].map(asGnTarget)
```

Comment:
what about `patches/*` files? these also need to be mapped.

Reply:
Can we assume that the file naming convention always maps to a existing file?

Reply:
just checked. We can parse the patch file using `git diff --numstat` it's what we do already to generate a patchinfo file. 

```
git apply --numstat -z brave/patches/base-test-BUILD.gn.patch
1       0       base/test/BUILD.gn
```

so we can get all the hunks via:

```
const {stdout} = await exec('git', ['--numstat', '-z', file]);
const affectedFiles = stdout.split('\n').map( hunk => hunk.split('\t').at(-1))
```

That works if a patch is added or modified.

However that doesn't cover patches removed. For that we probbaly need to retrieve the patch from the diff

Reply:
tempted to keep the bail logic for now and try to implement patchfile analysis in a seperate PR. But then I also wonder how often do we change patch files?

Reply:
that would be wise for now, yes. @goodov what are your thoughts about `chromium_src` and macro substitutions taking place in upstream sources in ways that we can't properly measure based on the file changed alone?

Reply:
I guess the dependency graph may be enough yeah.

Reply:
> how often do we change patch files?

often enough to support it similar to `chromium_src` files.

> what are your thoughts about `chromium_src` and macro substitutions taking place in upstream sources in ways that we can't properly measure based on the file changed alone?

we don't really care about changes inside, we only care about letting ninja/siso know that a GN-visible file's timestamp was changed, so scandeps and a possible rebuilt will be triggered.